File: cef-web/src/main/java/gov/epa/cef/web/service/ControlPathService.java
/*
 * © Copyright 2019 EPA CAERS Project Team
 *
 * This file is part of the Common Air Emissions Reporting System (CAERS).
 *
 * CAERS is free software: you can redistribute it and/or modify it under the 
 * terms of the GNU General Public License as published by the Free Software Foundation, 
 * either version 3 of the License, or (at your option) any later version.
 *
 * CAERS is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with CAERS.  If 
 * not, see <https://www.gnu.org/licenses/>.
*/
package gov.epa.cef.web.service;

import java.util.List;

import gov.epa.cef.web.service.dto.ControlAssignmentDto;
import gov.epa.cef.web.service.dto.ControlPathDto;
import gov.epa.cef.web.service.dto.ControlPathPollutantDto;
import gov.epa.cef.web.service.dto.bulkUpload.ControlPathBulkUploadDto;
import gov.epa.cef.web.service.dto.bulkUpload.ControlPathPollutantBulkUploadDto;

public interface ControlPathService {

	/**
     * Create a new Control Path
     * @param dto
     * @return
     */
	ControlPathDto create(ControlPathDto dto);

    /**
     * Retrieve Control Path by its id
     * @param id
     * @return
     */
    ControlPathDto retrieveById(Long id);

    /**
     * Retrieve Control Paths for a facility site
     * @param facilitySiteId
     * @return
     */
    List<ControlPathDto> retrieveForFacilitySite(Long facilitySiteId);
    
    /**
     * Retrieve Control Paths for an emissions process
     * @param processId
     * @return
     */
    List<ControlPathDto> retrieveForEmissionsProcess(Long processId);

    /**
     * Retrieve Control Paths for an emissions unit
     * @param unitId
     * @return
     */
    List<ControlPathDto> retrieveForEmissionsUnit(Long unitId);

    /**
     * Retrieve Control Paths for a release point
     * @param pointId
     * @return
     */
    List<ControlPathDto> retrieveForReleasePoint(Long pointId);
    
    /**
     * Retrieve Control Paths by control device id
     * @param controlDeviceId
     * @return
     */
	List<ControlPathDto> retrieveForControlDevice(Long controlDeviceId);

    /**
     * Update an existing Control Path by id
     * @param dto
     * @return
     */
    ControlPathDto update(ControlPathDto dto);
    
    /**
     * Retrieve parent Control Paths for a child Control Path
     * @param id
     * @return
     */
	List<ControlAssignmentDto> retrieveParentPathById(Long id);

    /**
     * Check if a control path was previously reported
     * @param controlPathId
     * @return
     */
    Boolean isPathPreviouslyReported(Long controlPathId);

    /**
     * Delete a Control Path for a given id
     * @param controlPathId
     */
    void delete(Long controlPathId);
    
    /**
     * Create a new Control Path Pollutant
     * @param dto
	 * @return 
     */
    ControlPathPollutantDto createPollutant(ControlPathPollutantDto dto);
    
    /**
     * Update existing Control Path Pollutant
     * @param dto
	 * @return 
     */
    ControlPathPollutantDto updateControlPathPollutant(ControlPathPollutantDto dto);
    
    /**
     * Delete a Control Path Pollutant for a given id
     * @param controlPathPollutantId
     */
    void deleteControlPathPollutant(Long controlPathPollutantId);
    	
	/**
     * Create a new Control Path Assignment
     * @param dto
	 * @return 
     */
    ControlAssignmentDto createAssignment(ControlAssignmentDto dto);
    
    /**
     * Retrieve Control Path Assignments by control path id
     * @param controlPathId
     * @return
     */
	List<ControlAssignmentDto> retrieveForControlPath(Long controlPathId);
	
    /**
     * Delete a Control Path Assignment for a given id
     * @param controlPathAssignmentId
     */
    void deleteAssignment(Long controlPathAssignmentId);
    
    /**
     * Update an existing Control Path Assignment by id
     * @param dto
     * @return
     */
    ControlAssignmentDto updateAssignment(ControlAssignmentDto dto);

    /**
     * Retrieve a list of control paths for the given program system code and emissions report year
     * @param programSystemCode
     * @param emissionsReportYear
     * @return
     */
    List<ControlPathBulkUploadDto> retrieveControlPaths(String programSystemCode, Short emissionsReportYear);

    /**
     * Retrieve a list of control path pollutants for the given program system code and emissions report year
     * @param programSystemCode
     * @param emissionsReportYear
     * @return
     */
    List<ControlPathPollutantBulkUploadDto> retrieveControlPathPollutants(String programSystemCode, Short emissionsReportYear);

}

File: cef-web/src/main/webapp/src/app/core/services/control-path.service.ts
/*
 * © Copyright 2019 EPA CAERS Project Team
 *
 * This file is part of the Common Air Emissions Reporting System (CAERS).
 *
 * CAERS is free software: you can redistribute it and/or modify it under the 
 * terms of the GNU General Public License as published by the Free Software Foundation, 
 * either version 3 of the License, or (at your option) any later version.
 *
 * CAERS is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with CAERS.  If 
 * not, see <https://www.gnu.org/licenses/>.
*/
import { Injectable } from '@angular/core';
import { ControlPath } from 'src/app/shared/models/control-path';
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { ControlAssignment } from 'src/app/shared/models/control-assignment';
import { ControlPathPollutant } from 'src/app/shared/models/control-path-pollutant';

@Injectable({
  providedIn: 'root'
})
export class ControlPathService {

  private baseUrl = 'api/controlPath';  // URL to web api

  constructor(private http: HttpClient) { }

  /** GET specified control path from the server */
  retrieve(id: number): Observable<ControlPath> {
    const url = `${this.baseUrl}/${id}`;
    return this.http.get<ControlPath>(url);
  }

  retrieveForEmissionsProcess(processId: number): Observable<ControlPath[]> {
    const url = `${this.baseUrl}/process/${processId}`;
    return this.http.get<ControlPath[]>(url);
  }

  retrieveForFacilitySite(facilitySiteId: number): Observable<ControlPath[]> {
    const url = `${this.baseUrl}/facilitySite/${facilitySiteId}`;
    return this.http.get<ControlPath[]>(url);
  }

  retrieveForEmissionsUnit(unitId: number): Observable<ControlPath[]> {
    const url = `${this.baseUrl}/unit/${unitId}`;
    return this.http.get<ControlPath[]>(url);
  }

  retrieveForReleasePoint(pointId: number): Observable<ControlPath[]> {
    const url = `${this.baseUrl}/releasePoint/${pointId}`;
    return this.http.get<ControlPath[]>(url);
  }

  retrieveForControlDevice(deviceId: number): Observable<ControlPath[]> {
    const url = `${this.baseUrl}/controlDevice/${deviceId}`;
    return this.http.get<ControlPath[]>(url);
  }

  update(controlPath: ControlPath): Observable<ControlPath> {
    const url = `${this.baseUrl}/${controlPath.id}`;
    return this.http.put<ControlPath>(url, controlPath);
  }

  isPathPreviouslyReported(id: number): Observable<boolean> {
    const url = `${this.baseUrl}/${id}/reported`;
    return this.http.get<boolean>(url);
  }

  delete(id: number): Observable<{}> {
    const url = `${this.baseUrl}/${id}`;
    return this.http.delete(url);
  }

  create(controlPath: ControlPath): Observable<ControlPath> {
    const url = `${this.baseUrl}`;
    return this.http.post<ControlPath>(url, controlPath);
  }

  retrieveAssignmentsForControlPath(controlPathId: number): Observable<ControlAssignment[]> {
    const url = `${this.baseUrl}/controlAssignment/${controlPathId}`;
    return this.http.get<ControlAssignment[]>(url);
  }

  retrieveParentAssignmentsForControlPathChild(controlPathId: number): Observable<ControlAssignment[]> {
    const url = `${this.baseUrl}/paretControlAssignment/${controlPathId}`;
    return this.http.get<ControlAssignment[]>(url);
  }

  createAssignmentForControlPath(controlAssignment: ControlAssignment): Observable<ControlAssignment> {
    const url = `${this.baseUrl}/controlAssignment/`;
    return this.http.post<ControlAssignment>(url, controlAssignment);
  }

  deleteAssignmentForControlPath(controlPathAssignmentId: number): Observable<{}> {
    const url = `${this.baseUrl}/controlAssignment/${controlPathAssignmentId}`;
    return this.http.delete(url);
  }

  updateAssignmentForControlPath(controlPathAssignment: ControlAssignment): Observable<ControlAssignment> {
    const url = `${this.baseUrl}/controlAssignment/${controlPathAssignment.id}`;
    return this.http.put<ControlAssignment>(url, controlPathAssignment);
  }

  /** Create Control Path Pollutant */
  createPollutant(controlPathPollutant: ControlPathPollutant): Observable<{}> {
    const url = `${this.baseUrl}/pollutant/`;
    return this.http.post<ControlPathPollutant>(url, controlPathPollutant);
  }

  /** Update Control Path Pollutant */
  updatePollutant(controlPathPollutant: ControlPathPollutant): Observable<{}> {
    const url = `${this.baseUrl}/pollutant/${controlPathPollutant.id}`;
    return this.http.put<ControlPathPollutant>(url, controlPathPollutant);
  }

  /** Delete Control Path Pollutant */
  deletePollutant(id: number): Observable<{}> {
    const url = `${this.baseUrl}/pollutant/${id}`;
    return this.http.delete<ControlPathPollutant>(url);
  }

}
